### PR TITLE
Adding CRUD Operations for DataSource

### PR DESCRIPTION
**Pull request by shekhar316**
## Description

This PR will do the following changes - 
- added support for adding the datasource
- added support for Hashmap of datasources

### Type of change

- [x] Refactoring Code
- [ ] New feature

## How has this been tested?

Using kruize-demo scripts. 
Image: `quay.io/rh-ee-shesaxen/autotune:kruize-crud4`

Verified database manually. 

**Test Configuration**
* Kubernetes clusters tested on: minikube

## Checklist :dart:

- [X] Followed coding guidelines
- [X] Comments added
- [X ] Dependent changes merged
- [ ] Documentation updated
- [ ] Tests added or updated

## Additional information
Logs: https://privatebin.corp.redhat.com/?1e3664cce5936ae4#HUaaU2fbMiZK2fuYuWrGmUzVTG2scaQxR2vtoG1CY1UC

### CODE REVIEW

**Comment by msvinaykumar** [comment]
U and D missing
Should we save it to DB ?

File: src/main/java/com/autotune/common/exceptions/DataSourceAlreadyExist.java
Comment (msvinaykumar) [inline src/main/java/com/autotune/common/exceptions/DataSourceAlreadyExist.java:1]: create sub folder under exception -> datasource and move all this exception there

File: tests/scripts/local_monitoring_tests/json_files/datasources.json [hunk 1]
```diff
@@ -0,0 +1,86 @@
+[
+    {
+        "name": "valid_without_serviceName",
```
Comment (msvinaykumar) [inline tests/scripts/local_monitoring_tests/json_files/datasources.json:3]: port ?

File: tests/scripts/local_monitoring_tests/json_files/datasources.json [hunk 2]
```diff
@@ -0,0 +1,86 @@
+[
+    {
```
Comment (msvinaykumar) [inline tests/scripts/local_monitoring_tests/json_files/datasources.json:2]: should we also have Cloud Provider name :   like AWS, EKS , Openshift etc ?

**Comment by shekhar316** [comment]
> U and D missing Should we save it to DB ?

We don't have APIs for U and D for now. I have written the back-end functions. 
And yes, we will save the data-sources to DB, but database part is not yet finalized, so I will adding database part for add datasource after PR#1151

File: src/main/java/com/autotune/common/exceptions/DataSourceAlreadyExist.java
Comment (shekhar316) [reply src/main/java/com/autotune/common/exceptions/DataSourceAlreadyExist.java:1]: Updated.

File: tests/scripts/local_monitoring_tests/json_files/datasources.json (hunk 2 above)
Comment (shekhar316) [reply tests/scripts/local_monitoring_tests/json_files/datasources.json:2]: Right now, we only support prometheus. Anything except that will result into UnSupportedDataSourceProvider exception.

**Review by msvinaykumar — APPROVED** [review]
LGTM

File: src/main/java/com/autotune/common/datasource/DataSourceCollection.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                // check the DB if the datasource already exists
+//                try {
+//                    DataSourceInfo dataSourceInfo = new ExperimentDBService().loadDataSourceFromDBByName(name);
+//                    if (null != dataSourceInfo) {
+//                        LOGGER.error("Datasource: {} already exists!", name);
+//                        continue;
+//                    }
+//                } catch (Exception e) {
+//                    LOGGER.error("Loading saved datasource {} failed: {} ", name, e.getMessage());
+//                }
```
Comment (dinogun) [inline src/main/java/com/autotune/common/datasource/DataSourceCollection.java:124]: In general we do not merge commented code, please remove

File: src/main/java/com/autotune/common/datasource/DataSourceCollection.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                // add the data source to DB
+                // TODO: add datasource to database
+                // addedToDB = new ExperimentDBService().addDataSourceToDB(datasource);
+                // if (addedToDB.isSuccess()) {
+                    LOGGER.info("Datasource added to the DB successfully.");
+                    // add to local map
+                    addDataSource(datasource);
+                // } else {
+                //    LOGGER.error("Failed to add datasource to DB: {}", addedToDB.getMessage());
+                // }
```
Comment (dinogun) [inline src/main/java/com/autotune/common/datasource/DataSourceCollection.java:147]: Commented code

File: src/main/java/com/autotune/common/datasource/DataSourceCollection.java (hunk 3 above)
Comment (shekhar316) [reply src/main/java/com/autotune/common/datasource/DataSourceCollection.java:124]: Updating the commented code for db operations according to recently merged db tables.

File: src/main/java/com/autotune/Autotune.java [hunk 5]
```diff
@@ -154,6 +161,29 @@ public static void main(String[] args) {
 
     }
 
+    /**
+     * Set up the data sources available at installation time from config file
+     */
+    private static void setUpDataSources() {
+        DataSourceCollection dataSourceCollection = DataSourceCollection.getInstance();
+        dataSourceCollection.addDataSourcesFromConfigFile(KruizeConstants.CONFIG_FILE);
```
Comment (dinogun) [inline src/main/java/com/autotune/Autotune.java:169]: Don't we have to read from the DB for all the existing the datasources at startup time? Remember that datasources can be added by the user after startup as well through the API

Comment (shekhar316) [reply src/main/java/com/autotune/Autotune.java:170]: @dinogun It's for the first time setup while adding datasources from YAML file. 

For fetching the data sources we can call `list_datasource` API added by shreya that will load all datasources from db.

Comment (shekhar316) [reply src/main/java/com/autotune/Autotune.java:170]: I think we can call `list_datasources` api while instantiating the dataSourceCollection object to make it sync with db.  Should I add it as a TODO item if it is okay?

Comment (shekhar316) [reply src/main/java/com/autotune/Autotune.java:172]: Added `loadDataSourcesFromDB()` to fetch available datasources from DB. 

**Kruize Pod Logs (On Initialization):** 
```java
2024-04-1711:50:51.666 INFO [main][KruizeHibernateUtil.java(63)]-DB is trying to connect to jdbc:postgresql://postgres-service:5432/kruizeDB
2024-04-1711:50:52.665 INFO [main][KruizeHibernateUtil.java(65)]-DB build session is successful !
2024-04-1711:50:52.792 INFO [main][Autotune.java(251)]-DB creation successful !
2024-04-1711:50:52.793 INFO [main][Autotune.java(259)]-DB Liveliness probe connection successful!
2024-04-1711:50:52.793 INFO [main][Autotune.java(109)]-Now running kruize local DDL's 
2024-04-1711:50:52.845 INFO [main][Autotune.java(251)]-DB creation successful !
2024-04-1711:50:52.846 INFO [main][Autotune.java(259)]-DB Liveliness probe connection successful!
2024-04-1711:50:52.847 INFO [main][DataSourceCollection.java(47)]-Checking available datasources from database:
2024-04-1711:50:53.005 INFO [main][DataSourceCollection.java(50)]-No datasource found in database.
2024-04-1711:50:53.044 INFO [main][DataSourceCollection.java(89)]-Trying to add the datasource to collection: prometheus-1
2024-04-1711:50:53.045 INFO [main][DataSourceCollection.java(97)]-Verifying datasource reachability status: prometheus-1
2024-04-1711:50:53.201 INFO [main][DataSourceCollection.java(100)]-Datasource is serviceable.
2024-04-1711:50:53.234 INFO [main][DataSourceCollection.java(104)]-Datasource added to the DB successfully.
2024-04-1711:50:53.234 INFO [main][DataSourceCollection.java(109)]-Datasource added to the collection successfully.
2024-04-1711:50:53.234 INFO [main][Autotune.java(188)]-Checking available datasources:
2024-04-1711:50:53.234 INFO [main][Autotune.java(194)]-Datasource found: prometheus-1, http://prometheus-k8s.monitoring.svc.cluster.local:9090
2024-04-1711:50:53.238 INFO [main][KruizeHibernateUtil.java(63)]-DB is trying to connect to jdbc:postgresql://postgres-service:5432/kruizeDB
2024-04-1711:50:53.338 INFO [main][KruizeHibernateUtil.java(65)]-DB build session is successful !
```


**Kruize Pod Logs (On Restart):** 
```java
2024-04-1711:51:23.966 INFO [main][KruizeHibernateUtil.java(63)]-DB is trying to connect to jdbc:postgresql://postgres-service:5432/kruizeDB
2024-04-1711:51:24.864 INFO [main][KruizeHibernateUtil.java(65)]-DB build session is successful !
2024-04-1711:51:24.931 WARN [main][Autotune.java(240)]-sql: alter table if exists kruize_experiments add constraint UK_experiment_name unique (experiment_name); failed due to : Converting `org.hibernate.exception.SQLGrammarException` to JPA `PersistenceException` : JDBC exception executing SQL [alter table if exists kruize_experiments add constraint UK_experiment_name unique (experiment_name);]
2024-04-1711:51:24.936 INFO [main][Autotune.java(251)]-DB creation successful !
2024-04-1711:51:24.937 INFO [main][Autotune.java(259)]-DB Liveliness probe connection successful!
2024-04-1711:51:24.937 INFO [main][Autotune.java(109)]-Now running kruize local DDL's 
2024-04-1711:51:24.940 WARN [main][Autotune.java(240)]-sql: alter table kruize_experiments add column metadata_id bigint references kruize_dsmetadata(id), alter column datasource type varchar(255); failed due to : Converting `org.hibernate.exception.SQLGrammarException` to JPA `PersistenceException` : JDBC exception executing SQL [alter table kruize_experiments add column metadata_id bigint references kruize_dsmetadata(id), alter column datasource type varchar(255);]
2024-04-1711:51:24.941 INFO [main][Autotune.java(251)]-DB creation successful !
2024-04-1711:51:24.941 INFO [main][Autotune.java(259)]-DB Liveliness probe connection successful!
2024-04-1711:51:24.942 INFO [main][DataSourceCollection.java(47)]-Checking available datasources from database:
2024-04-1711:51:25.136 INFO [main][DataSourceCollection.java(53)]-Datasource found: prometheus-1
2024-04-1711:51:25.179 ERROR [main][DataSourceCollection.java(146)]-Datasource: prometheus-1 already exists!
2024-04-1711:51:25.180 INFO [main][Autotune.java(188)]-Checking available datasources:
2024-04-1711:51:25.180 INFO [main][Autotune.java(194)]-Datasource found: prometheus-1, http://prometheus-k8s.monitoring.svc.cluster.local:9090
2024-04-1711:51:25.184 INFO [main][KruizeHibernateUtil.java(63)]-DB is trying to connect to jdbc:postgresql://postgres-service:5432/kruizeDB
2024-04-1711:51:25.273 INFO [main][KruizeHibernateUtil.java(65)]-DB build session is successful !
```

Comment (shekhar316) [reply src/main/java/com/autotune/Autotune.java:172]: **Update Logs After Restart (With Updated Error Msgs):** 
```java
2024-04-1712:39:53.305 INFO [main][KruizeHibernateUtil.java(63)]-DB is trying to connect to jdbc:postgresql://postgres-service:5432/kruizeDB
2024-04-1712:39:54.239 INFO [main][KruizeHibernateUtil.java(65)]-DB build session is successful !
2024-04-1712:39:54.302 WARN [main][Autotune.java(240)]-sql: alter table if exists kruize_experiments add constraint UK_experiment_name unique (experiment_name); failed due to : add constraint - Attempted operation has already been executed
2024-04-1712:39:54.308 INFO [main][Autotune.java(253)]-DB creation successful !
2024-04-1712:39:54.309 INFO [main][Autotune.java(261)]-DB Liveliness probe connection successful!
2024-04-1712:39:54.309 INFO [main][Autotune.java(109)]-Now running kruize local DDL's 
2024-04-1712:39:54.312 WARN [main][Autotune.java(242)]-sql: alter table kruize_experiments add column metadata_id bigint references kruize_dsmetadata(id), alter column datasource type varchar(255); failed due to : add column - Attempted operation has already been executed
2024-04-1712:39:54.313 INFO [main][Autotune.java(253)]-DB creation successful !
2024-04-1712:39:54.313 INFO [main][Autotune.java(261)]-DB Liveliness probe connection successful!
2024-04-1712:39:54.315 INFO [main][DataSourceCollection.java(47)]-Checking available datasources from database:
2024-04-1712:39:54.500 INFO [main][DataSourceCollection.java(53)]-Datasource found: prometheus-1
2024-04-1712:39:54.541 ERROR [main][DataSourceCollection.java(146)]-Datasource: prometheus-1 already exists!
2024-04-1712:39:54.542 INFO [main][Autotune.java(188)]-Checking available datasources:
2024-04-1712:39:54.542 INFO [main][Autotune.java(194)]-Datasource found: prometheus-1, http://prometheus-k8s.monitoring.svc.cluster.local:9090
2024-04-1712:39:54.545 INFO [main][KruizeHibernateUtil.java(63)]-DB is trying to connect to jdbc:postgresql://postgres-service:5432/kruizeDB
2024-04-1712:39:54.630 INFO [main][KruizeHibernateUtil.java(65)]-DB build session is successful !

```

**Review by dinogun — APPROVED** [review]
LGTM

**Pull request merged**